gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/visitor/ExecutionVisitor.java

```diff
@@ -141,53 +144,83 @@ public EntityFetcherResponse visit(SelectionNode selectionNode) {
     Filter filter = constructFilterFromChildNodesResult(result);
-    // Select attributes, metric aggregations and timeseries data from corresponding sources
-    List<EntityFetcherResponse> resultMapList =
-        selectionNode.getAttrSelectionSources().parallelStream()
-            .map(
-                source -> {
-                  EntitiesRequest request =
-                      EntitiesRequest.newBuilder(executionContext.getEntitiesRequest())
-                          .clearSelection()
-                          .clearFilter()
-                          .addAllSelection(
-                              executionContext.getSourceToSelectionExpressionMap().get(source))
-                          .setFilter(filter)
-                          .build();
-                  IEntityFetcher entityFetcher = queryHandlerRegistry.getEntityFetcher(source);
-                  EntitiesRequestContext context =
-                      new EntitiesRequestContext(
-                          executionContext.getTenantId(),
-                          request.getStartTimeMillis(),
-                          request.getEndTimeMillis(),
-                          request.getEntityType(),
-                          executionContext.getRequestHeaders());
-                  return entityFetcher.getEntities(context, request);
-                })
-            .collect(Collectors.toList());
-    resultMapList.addAll(
-        selectionNode.getAggMetricSelectionSources().parallelStream()
-            .map(
-                source -> {
-                  EntitiesRequest request =
-                      EntitiesRequest.newBuilder(executionContext.getEntitiesRequest())
-                          .clearSelection()
-                          .clearFilter()
-                          .addAllSelection(
-                              executionContext.getSourceToMetricExpressionMap().get(source))
-                          .setFilter(filter)
-                          .build();
-                  IEntityFetcher entityFetcher = queryHandlerRegistry.getEntityFetcher(source);
-                  EntitiesRequestContext context =
-                      new EntitiesRequestContext(
-                          executionContext.getTenantId(),
-                          request.getStartTimeMillis(),
-                          request.getEndTimeMillis(),
-                          request.getEntityType(),
-                          executionContext.getRequestHeaders());
-                  return entityFetcher.getAggregatedMetrics(context, request);
-                })
-            .collect(Collectors.toList()));
+    // Select attributes, metric aggregations and time-series data from corresponding sources
+    List<EntityFetcherResponse> resultMapList = new ArrayList<>();
+    // if single-sourced and same source, then combine get Entities and aggregated Metrics
+    if (isSingleSourceAndSame(selectionNode.getAttrSelectionSources(), selectionNode.getAggMetricSelectionSources())) {
```

Comment:
The ExecutionVisitor should ideally not be aware of the single vs multiple source. All that complex logic should just be in one place and that is in the ExecutionTreeBuilder. 
The decision to use getEntitiesAndAggregatedMetrics  should purely be based on whether the the SelectionNode has both AttrSelectionSources and/or AggMetricSelectionSources. 
That way the code will be very generic.

cc: @tim-mwangi Please see if that is possible and cleanup if you can.

Reply:
Ahhh... sorry... I take that comment back. I misunderstood this a bit.
Please ignore my comment.

Reply:
It's possible though if you wanna do it that way, you will have to introduce more QueryNode types. For example EDS does not support aggregations or time aggregations and to support all same source type of queries you will have to either branch in execution visitor or use a new node type if you don't want to branch.